### ISSUE TITLE
Maximum length of strip? 

### ISSUE THREAD

**Issue by mzeile** [adammhaile/RPi-LPD8806#11]
Hello and thanks for your library! 

I'm about to order 25 meters of Lpd8806 52 pixel/m strips for my living room and I'm wondering myself if such a length can be addressed using your library? What's the limit there with one raspberry? (power supply every 5 meters to the strip is available of course) 

Thanks for help! 
Michael

**Comment by adammhaile** [comment]
WOW! Sounds like a huge project! :) I HAVE to see some pictures when you
are done :) Sounds awesome.
So... I have an install on my front port that is 10 meters of the 36/m
strips, which works quite nicely, except that they definitely run slowly.
I'm maxing out at around 10 frames per second. You are talking about nearly
4 times that many pixels. So, it _might_ work (I'm not even sure the data
signal can go that far) but you'd probably get, at best, 2-3 frames per
second. So, would that even be acceptable? I'm not sure what effect you are
going for. To get any good performance (the bottleneck will absolutely be
the SPI interface) you will likely need multiple Pis (at least 2, if not
more).

Sadly, my library doesn't exactly support for synchronizing multiple Pis.
At the moment. I've been working on some updates to it that completely
detach the animation generation for the actual updating of the strip. The
intent was to allow multiple "drivers" that would support strips _other_ that
LPD8806. But, I've also been toying around with having the animation data
sent over the network to a receiver that would then update the strips. It's
all very conceptual at the moment. But, I'd be happy to work with you on
it. Sounds like a fun project.

**Comment by mzeile** [comment]
Thanks for your input! I'm not trying to create a disco in my living room... but more than 2-3 updates/second would be welcome! I've asked about this kind of issue in the FastLED github - I hope that someone there can give some more details about long strips and high updates rates. You're saying the SPI interface is the bottleneck... then it may really be worth looking into something else - with 3-wire WS2811 / WS2812 LED's even if that requires me to go away from the Raspberry Pi platform. But I need to be sure before I'm going to order 25 meters! ;)

**Comment by adammhaile** [comment]
Yeah, definitely the one major downside to the LPD8806 strips and the RPi.
Note, that for the WS2812 strips, you won't be able to use the RPi easily,
at least not for that many. They are definitely faster, but I can't really
tell you by how much.

**Comment by focalintent** [comment]
Just as a side note - the WS2812's aren't faster - they're 800kbps data rates, while the LPD8806 can run at a 20Mhz SPI clock rate (though I don't know what the rPI's max SPI data rate is).

**Comment by adammhaile** [comment]
Good to know. I've clocked the LPD8806 at up to 12MHz before I saw
problems, but it seems the bottleneck must be further up the chain... or my
code is crap (this is possible :P )

**Comment by focalintent** [comment]
I think it depends on the strip length and the strength of the data signal coming out of your device, since the LPD8806 doesn't regenerate the signal at each chip (at least, I believe it doesn't).  I've pushed the LPD8806's at 24Mhz clocks before, but yeah, the longer the strip the lower the clock speed I need to hit it with.  Good to hear that 12Mhz is still doable though!

I'm not sure where this idea that the WS2811 and friends are "faster" came from - it's an 800kbps data rate, and a fairly rigid one at that.  Libraries like OctoWS2811 and FadeCandy help by allowing you to drive 8 lines in parallel, but even then, you're talking about 6.4Mbps data rates, anywhere from a half to a quarter of what you can get with the LPD8806.  (That said, I love the WS2812B's packaging - 4 pin led w/WS2812 hardware embedded?  Makes for far more flexible strips/layouts, not to mention density - i've see 144 led/meter WS2812 strips out there!)

**Comment by mzeile** [comment]
Thanks for all your comments. I'm not still much further yet. Did really no one try to add 10+ or 20+ meters together? ;) I've found the data sheet of the LPD8806 here:

http://www.gree-leds.com/newsshow.asp?showidd=352

It seems that it does regenerate the signal: "Unique data clock regeneration mechanism, super signal drive capability, support cascading length over 2000 pixels."

I think I'll have to do the trial and error approach... at least I know that I can access 10 meters for sure. For everything further I'll have to see what happens. I'll report back to you guys in some months.

**Comment by adammhaile** [comment]
I have done 10 meters and it works fine. You should be OK for more and I
believe you are right that the LPD8806 chip cascades the signal... it's
kinda the point of the chip.
HOWEVER, you will still have a speed issue if run directly from the Pi.
I've seen some fancy tricks to run it faster on HackADay.com, but can't
vouch for any of those.
One option is that the Pi actually has 2 SPI ports so you might be able to
break it up. Or run two Pis :P

**Comment by adammhaile** [comment]
mzeile... how's that project going?
So... while it will still not be FAST by any means I am working on an overhaul to the library that might help you out and I would love a tester ;)
I have reworked the SPI driver so that I have increased my performance with actually sending the data to the strip (not the animation generation) by 5x. It used to take 100+ ms to send data to 360 pixels and now I'm doing almost 600 in < 20ms. Granted, BY FAR the biggest bottleneck I have seen is actually generating each frame. The Pi kind of is horrible at float math, and I'm looking to move to the BeagleBone Black for my projects with more lights... which this library _should_ support.

Let me know and I'll get you an alpha version of my new library. I'm not doing the development on it in this repro because it's a complete rework that will support not only more than the Rasp Pi as a host (to include Beagle Bone and even Mac/PC/Linux) but will support just about any LED strip or display.

**Comment by focalintent** [comment]
Another thing to look into doing is ditch the floating point math.  For all the stuff that I do using FastLED I use fixed point math with 16 and 32 bit integers.  At its simplest, a 16 bit integer can represent an 8 bit integer and an 8 bit "fractional" amount.  Or you can represent just a fractional amount.  I just made very very heavy use of this to do a stupidly fast perlin simplex noise implementation on AVR that can do 30k points of simplex noise/second (where most people say "just generate a couple of points of noise then interpolate the rest of your pixels between them").

**Comment by adammhaile** [comment]
Nice library. Was unaware of that one... totally going to use that for my
Arduino stuff!

I've actually been looking into ditching the float altogether... I'll
definitely take a look at your code and see if I can figure out how to do
that in Python.
Fortunately, for the most part, I don't use floats... it's mainly in the
brightness calculations (for master brightness). But those get run on every
frame output on every pixel, so it's not great... partly I think the
problem is just python. I've trying running some of this with PyPy and,
while it's not super stable, it's wicked fast. Which generally tells me
that my bottleneck is python itself... but I'll take whatever speed boosts
I can get!

**Comment by focalintent** [comment]
Here's an example of fixed point math for brightness scaling that I use in FastLED.  Represent brightness as a scale from 0-255.  Now, for each of the r,g,b values you can say:

```
 uint8_t val = (r * scale) >> 8;
```

so that's a multiply, and an 8 bit shift to the right (which on avr we have some asm to take advantage of the fact that we're really just copying the top 8bit register of the result).  

In fact, this is fast enough that when I'm using hardware SPI to push data out, I have more than enough cycles to scale the byte i'm about to send out while waiting for the hardware SPI subsystem to finish sending out the previous byte - effectively giving brightness scaling "for free" in FastLED.  (I do something similar with interleaving operations in between toggling pins for WS2812 output to get it for free there as well - you don't want to see what I have to do on platforms that don't have a hardware multiply).

**Comment by adammhaile** [comment]
Nice! Thanks! Definitely going to use that :)

**Comment by mzeile** [comment]
Hello adammhaile,
sorry but I was busy with other stuff so I could not invest more time in my project. Back in April I did a few tests. One with this library:
https://github.com/fishkingsin/elinux-lpd8806
and with this one:
https://www.npmjs.org/package/lpd8806

All I can say is that the refresh rate with the c-library was >10 times faster and was using 95% less CPU of my Pi :) I've made two videos back then which may help you to judge what update rates I'm taking about:
http://home.primusnetz.de/mzeile/led/

The only thing I changed was in 'lpd8806led.c' where I adjusted the SPI clock speed down from the defined 15 MHZ. The longer the strip the less MHZ. For my 20 meters test I _think_ I was using something around 3-4 MHZ?!

So for my project (I'll continue later this year) I'll go with this tiny elinux-lpd8806 library since it's my favorite c-language and also it seems to have the SPI bus as the only bottleneck (the 'rainbow' app was running with less than 1% CPU usage).
